### PR TITLE
[WIP] Dynamic View Controller

### PR DESCRIPTION
**Pull request by descorp**
### Changes

* add FoldableFormItem
* add DynamicViewController
* propagate DynamicViewController up to DropIn presentation level
* some minor refactoring

### Unrelated changes

* diners regex update
* AVS check for issuing country

### CODE REVIEW

File: Adyen/UI/Form/FormItemViewBuilder.swift [hunk 1]
```diff
@@ -59,4 +59,16 @@ public struct FormItemViewBuilder {
     public func build(with item: FormSeparatorItem) -> FormItemView<FormSeparatorItem> {
         FormSeparatorItemView(item: item)
     }
+
+    /// Builds `FormSeparatorItemView` from `FormSeparatorItem`.
+    /// :nodoc:
+    public func build(with item: FormFoldableItem) -> FormItemView<FormFoldableItem> {
+        FormFoldableItemView(item: item)
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/FormItemViewBuilder.swift:66]: I prefer `FoldableFormItemView`, the parent class name should have the subclass qualifier at the beginning:

```
parent class: FormItemView
subclass: FoldableFormItemView
```

```
parent class: ParentClass
subclass: ChildParentClass
```

File: Adyen/UI/Form/Items/Foldable/FormFoldableItem.swift [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// Copyright (c) 2021 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import Foundation
+
+/// A form item that presents single FormItem, that could be hide or showed dynamicaly.
+/// :nodoc:
+public final class FormFoldableItem: CompoundFormItem {
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Foldable/FormFoldableItem.swift:11]: same as [here](https://github.com/Adyen/adyen-ios/pull/442/files#r605736167)

File: Adyen/UI/Form/Items/Split/FormSplitItemView.swift [hunk 3]
```diff
@@ -9,39 +9,24 @@ import UIKit
 
 /// A view representing a split item.
 internal final class FormSplitItemView: FormItemView<FormSplitItem> {
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Split/FormSplitItemView.swift:11]: `SplitFormItemView` sounds better IMO

Comment (descorp) [reply Adyen/UI/Form/Items/Split/FormSplitItemView.swift:11]: Everything is called FormBlablaItem. 
I believe those should align 🤔

File: Adyen/UI/Form/FormItemViewBuilder.swift (hunk 1 above)
Comment (descorp) [reply Adyen/UI/Form/FormItemViewBuilder.swift:66]: We can rename, but this would be a breaking change.
Let's address it in another PR

File: Adyen/UI/Form/Items/Split/FormSplitItemView.swift (hunk 3 above)
Comment (mohammedDehairy) [reply Adyen/UI/Form/Items/Split/FormSplitItemView.swift:11]: yeah, true, but I prefer the other way tbh

File: Adyen/UI/Form/FormViewController.swift [hunk 4]
```diff
@@ -198,17 +154,6 @@ open class FormViewController: UIViewController, Localizable {
...
-        guard let keyboardFrame = notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect else { return }
-        var heightOffset = keyboardFrame.origin.y - UIScreen.main.bounds.height
-        
-        if #available(iOS 11.0, *) {
-            heightOffset += min(abs(heightOffset), view.safeAreaInsets.bottom)
-        }
-        
-        bottomConstraint?.constant = heightOffset
-    }
-
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/FormViewController.swift:211]: why is this removed?

File: Adyen/UI/UI Style/CoreFonts.swift [hunk 5]
```diff
@@ -17,11 +17,7 @@ extension UIFont {
         }
         
         internal static var largeTitle: UIFont {
-            if #available(iOS 11.0, *) {
-                return UIFont.preferredFont(forTextStyle: .largeTitle).adyen.font(with: .bold)
-            } else {
-                return .systemFont(ofSize: 32.0, weight: .bold)
-            }
+            .systemFont(ofSize: 32.0, weight: .bold)
```
Comment (mohammedDehairy) [inline Adyen/UI/UI Style/CoreFonts.swift:20]: this should be:
```
 UIFont.preferredFont(forTextStyle: .largeTitle).adyen.font(with: .bold)
```

File: Adyen/UI/UI Style/CoreColors.swift [hunk 6]
```diff
@@ -82,13 +82,7 @@ extension UIColor {
 
         /// :nodoc:
         public static var componentLoadingMessageColor: UIColor {
-            if #available(iOS 11.0, *) {
-                return UIColor(named: "awaitLoadingMessageColor",
-                               in: Bundle.coreInternalResources,
-                               compatibleWith: nil) ?? componentPlaceholderText
-            } else {
-                return componentPlaceholderText
-            }
```
Comment (mohammedDehairy) [inline Adyen/UI/UI Style/CoreColors.swift:91]: this should be:
```
UIColor(named: "awaitLoadingMessageColor",
                                in: Bundle.coreInternalResources,
                                compatibleWith: nil) ?? componentPlaceholderText
```

File: Adyen/UI/View Controllers/ViewControllerDelegate.swift [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+            objc_getAssociatedObject(self, &AdyenViewControllerAssociatedKeys.dynamicContentDelegate) as? DynamicViewControllerDelegate
+        }
+        set {
+            objc_setAssociatedObject(self,
+                                     &AdyenViewControllerAssociatedKeys.dynamicContentDelegate,
+                                     newValue,
+                                     objc_AssociationPolicy.OBJC_ASSOCIATION_ASSIGN)
+        }
+    }
+
```
Comment (mohammedDehairy) [inline Adyen/UI/View Controllers/ViewControllerDelegate.swift:56]: we use this pattern a lot, it makes me uncomfortable, its not the swift way, what's wrong in implementing a delegate property in all conforming classes?

File: AdyenCard/Utilities/CardType/CardBrandProvider.swift [hunk 8]
```diff
@@ -26,7 +26,7 @@ internal struct CardBrandProviderParameters {
 /// Provide cardType detection based on BinLookup API.
 internal final class CardBrandProvider: AnyCardBrandProvider {
     
-    private static let minBinLength = 6
+    private static let minBinLength = 10
```
Comment (mohammedDehairy) [inline AdyenCard/Utilities/CardType/CardBrandProvider.swift:29]: according to Nicholas its 11 actually

File: Adyen/UI/Form/FormViewController.swift (hunk 4 above)
Comment (descorp) [reply Adyen/UI/Form/FormViewController.swift:211]: I am still hovering on this one as well..

IN one hand: 
We have too many sources of trust for keyboard events. 
Too many things happening in the same time in different parts of the app and it is not easy to control animation.

In the other:
This will lead to changes in Component integration. Also previously we had a white "background" under keyboard, I will have to do something extra.

File: Adyen/UI/UI Style/CoreFonts.swift (hunk 5 above)
Comment (descorp) [reply Adyen/UI/UI Style/CoreFonts.swift:20]: oh, good one!

File: Adyen/UI/UI Style/CoreColors.swift (hunk 6 above)
Comment (descorp) [reply Adyen/UI/UI Style/CoreColors.swift:91]: 👍

File: Adyen/UI/View Controllers/ViewControllerDelegate.swift (hunk 7 above)
Comment (descorp) [reply Adyen/UI/View Controllers/ViewControllerDelegate.swift:56]: good point!

File: AdyenCard/Utilities/CardType/CardBrandProvider.swift (hunk 8 above)
Comment (descorp) [reply AdyenCard/Utilities/CardType/CardBrandProvider.swift:29]: Yep. We do `bin > minBinLength ` :D

Comment (mohammedDehairy) [reply AdyenCard/Utilities/CardType/CardBrandProvider.swift:29]: minimum means its an allowed value, so if the minimum is 10 then it should be `if bin >= minBinLength ` :-D

Comment (descorp) [reply AdyenCard/Utilities/CardType/CardBrandProvider.swift:29]: 👍
